From 18181557ffb6a186d358e967b279cd6c4d91f965 Mon Sep 17 00:00:00 2001
From: SagarMakhar <sagarmakhar@gmail.com>
Date: Tue, 5 Oct 2021 04:53:35 +0000
Subject: [PATCH] Ignore netd errors for < 4.9 kernels

Change-Id: I5f1296bf6e677dda05eceb6a480d7d669ef556b6
Signed-off-by: SagarMakhar <sagarmakhar@gmail.com>
(cherry picked from commit e3c9266d2c149f02704268c3a5a0fe1f78b5ae3b)
(cherry picked from commit 83e0a9454a9b34b99655a930cbd43915760b9fa1)
---
 server/Controllers.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/server/Controllers.cpp b/server/Controllers.cpp
index 1f2bac22..91c2f39f 100644
--- a/server/Controllers.cpp
+++ b/server/Controllers.cpp
@@ -19,6 +19,7 @@
 #include <set>
 #include <string>
 
+#include <android-base/properties.h>
 #include <android-base/stringprintf.h>
 #include <android-base/strings.h>
 #include <netdutils/Stopwatch.h>
@@ -276,6 +277,8 @@ void Controllers::initIptablesRules() {
 }
 
 void Controllers::init() {
+    bool ebpf_supported = android::base::GetBoolProperty("ro.kernel.ebpf.supported", true);
+
     initIptablesRules();
     Stopwatch s;
 
@@ -283,7 +286,7 @@ void Controllers::init() {
     gLog.info("Initializing ClatdController: %" PRId64 "us", s.getTimeAndResetUs());
 
     netdutils::Status tcStatus = trafficCtrl.start();
-    if (!isOk(tcStatus)) {
+    if (!isOk(tcStatus) && ebpf_supported) {
         gLog.error("Failed to start trafficcontroller: (%s)", toString(tcStatus).c_str());
         gLog.error("CRITICAL: sleeping 60 seconds, netd exiting with failure, crash loop likely!");
         // The expected reason we get here is a major kernel or other code bug, as such
